test/gtest/stringify.cpp

```diff
@@ -1,151 +1,699 @@
+
+  void addUniqueSymbol(SeparatorReason curr) {
+    if (auto* func = curr.getFuncStart()) {
+      visited.push_back(FuncStart{func->func});
+    } else if (auto* block = curr.getBlockStart()) {
+      visited.push_back(BlockStart{block->block});
+    } else if (auto* iff = curr.getIfStart()) {
+      visited.push_back(IfStart{iff->iff});
+    } else if (curr.getElseStart()) {
+      visited.push_back(ElseStart{});
+    } else if (auto* loop = curr.getLoopStart()) {
+      visited.push_back(LoopStart{loop->loop});
+    } else if (auto* tryy = curr.getTryStart()) {
+      visited.push_back(TryStart{tryy->tryy});
+    } else if (auto* tag = curr.getCatchStart()) {
+      visited.push_back(CatchStart{tag->tag});
+    } else if (curr.getCatchAllStart()) {
+      visited.push_back(CatchAllStart{});
+    } else if (auto* tryt = curr.getTryTableStart()) {
+      visited.push_back(TryTableStart{tryt->tryt});
+    } else if (auto* end = curr.getEnd()) {
+      visited.push_back(End{end->curr});
     }
-  };
+  }
+};
+
+TEST(StringifyWalkerTest, Nop) {
+  Module wasm;
+  Builder builder(wasm);
+
+  // (func $func nop)
+  auto* nop = builder.makeNop();
+  auto f = builder.makeFunction("func", Signature(), {}, nop);
+  auto* func = f.get();
+
+  wasm.addFunction(std::move(f));
+  TestStringifyWalker walker;
+  walker.walkModule(&wasm);
+
+  std::vector<Visited> expected{
+    Visited(FuncStart{func}), Visited(nop), Visited(End{nullptr})};
+  EXPECT_EQ(walker.visited, expected);
+}
+
+TEST(StringifyWalkerTest, Add) {
+  Module wasm;
+  Builder builder(wasm);
+
+  // (func $func
+  //   (i32.add
+  //     (i32.const 0)
+  //     (i32.const 1)
+  //   )
+  // )
+  auto* a = builder.makeConst(uint32_t(0));
+  auto* b = builder.makeConst(uint32_t(1));
+  auto* add = builder.makeBinary(AddInt32, a, b);
+  auto f = builder.makeFunction("func", Signature(), {}, add);
+  auto* func = f.get();
+
+  wasm.addFunction(std::move(f));
+  TestStringifyWalker walker;
+  walker.walkModule(&wasm);
+
+  std::vector<Visited> expected{Visited(FuncStart{func}),
+                                Visited(a),
+                                Visited(b),
+                                Visited(add),
+                                Visited(End{nullptr})};
+  EXPECT_EQ(walker.visited, expected);
+}
 
+TEST(StringifyWalkerTest, EmptyBlock) {
   Module wasm;
-  parseWast(wasm, moduleText);
+  Builder builder(wasm);
+
+  // (func $func
+  //   (block)
+  // )
+  auto* block = builder.makeBlock();
+  auto f = builder.makeFunction("func", Signature(), {}, block);
+  auto* func = f.get();
+
+  wasm.addFunction(std::move(f));
+  TestStringifyWalker walker;
+  walker.walkModule(&wasm);
+
+  std::vector<Visited> expected{Visited(FuncStart{func}),
+                                Visited(block),
+                                Visited(End{nullptr}),
+                                Visited(BlockStart{block}),
+                                Visited(End{nullptr})};
```

Comment:
Why are there two blocks here? If one is an implicit block, why aren't they nested?

Reply:
The function body counts as its own control flow structure, then the single block counts as the one additional control flow structure. By implicit block, do you mean an unnamed block? We don't treat those specially.

Reply:
The whole point of StringifyWalker is to un-nest control flow :) This lets us treat control flow structures as single "characters" in the substring describing their parents, and then later have another substring describing their contents. Control flow delimiters, including block starts and ends, are emitted as globally unique characters, so they never show up in repeated substrings. The un-nesting combined with the unique delimiters allows us to outline sequences containing control flow structures but prevents us from trying to outline e.g. an `end` or an `else` without the rest of its structure.

Reply:
Hmm, I guess I'm confused as to how we can keep track of what appears where, if we entirely un-nest like this... but if it's a pre-existing issue of my understanding, let's not block this on that.

Reply:
We never actually un-nest the IR, and each start delimiter contains a pointer to the expression it's starting, so when we reconstruct the IR we know what control flow structure we are reconstructing the contents of and its relationship with its parent is never changed.

Reply:
But it seems to be a BFS on control flow while normal otherwise? The next test shows `block, end, block, nop, end` where the nop is in the right place, inside its block. But that means that code like this
```
(block $A
  (block $B
  )
  (nop)
  (block $C)
)
```
would be `block,nop,end,block,end,block,end`. So the blocks are entirely out of order wrt the nop. Is that right?

Reply:
Yes, exactly. Everything that is not a control flow structure is scanned in the normal postorder so we get the sequence of instructions that will appear inside each control flow structure just as it will be in the binary format. I just added your example as a new test. Note that blocks appear as individual entries in the correct locations with respect to their siblings, and they also appear later in un-nested `StartBlock` delimiters followed by their contents.

Reply:
Note that each control flow structure shows up _twice_ in the visitation order. The first time it is as a single unit inside its parent and ordered properly with its siblings. This is the appearance that allows the control flow structure to appear in a common substring and get outlined. The second time it appears as a parameter inside a start delimiter followed by all its children. This is the appearance that allows us to associate the control flow structure with its children when we reconstruct the IR.